internal/controller/postgrescluster/patroni.go

```diff
@@ -499,6 +512,44 @@ func (r *Reconciler) reconcilePatroniSwitchover(ctx context.Context,
 			stdout, stderr, command...)
 	}
 
+	// To ensure idempotency, the operator verifies that the timeline reported by Patroni
+	// matches the timeline that was present when the switchover was first requested.
+	// TODO(benjb): consider pulling the timeline from the pod annotation; manual experiments
+	// have shown that the annotation on the Leader pod is up to date during a switchover, but
+	// missing from the Replica pods.
+	timeline, err := patroni.Executor(exec).GetTimeline(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	if timeline == 0 {
+		return errors.New("error getting and parsing current timeline")
+	}
+
+	statusTimeline := cluster.Status.Patroni.ExpectedTimeline
+
+	// If the `ExpectedTimeline` field is empty, this is the first reconcile after
+	// a switchover has been requested and we need to fill in the field with the current TL
+	// as reported by Patroni.
+	// We return from here without calling for an explicit requeue, but since we're updating
+	// the object, we will reconcile this again for the actual switchover/failover action.
+	if statusTimeline == nil || (statusTimeline != nil && *statusTimeline == 0) {
+		log.V(1).Info("Setting ExpectedTimeline to current timeline")
+		cluster.Status.Patroni.ExpectedTimeline = &timeline
+		return err
+	}
+
+	// If the `ExpectedTimeline` field does not match the current timeline as reported by Patroni,
+	// then we assume a switchover has been completed, and we have reached this point because the
```

Comment:
:thinking: I'm not sure I'm following why it's safe to make this assumption (although I'm not saying it isn't!). I'm mainly curious if there could be a scenario where the current timeline doesn't match but the switchover has not completed.